### PR TITLE
dejadup init:34.2. 

### PR DESCRIPTION
**Pull request by s1lvester**
###### Motivation for this change

See Issue: #17763 
###### Things done
- [ ] Tested using sandboxing
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/backup/dejadup/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{ stdenv, fetchurl, cmake, pkgconfig, vala, 
+  gettext, gnome3, libnotify, intltool, itstool, 
+  glib, gtk3, packagekit, libsecret, ... }:
```
Comment (abbradar) [inline pkgs/applications/backup/dejadup/default.nix]: You usually don't need to add a catch-all (`...`).

**Comment by abbradar** [comment]
I think you can fix the issue by adding `vala` to `buildInputs` of `libsecret`.

**Comment by RamKromberg** [comment]
> I think you can fix the issue by adding vala to buildInputs of libsecret.

I'm afraid it's a bit more involved then that: https://developer.gnome.org/libsecret/unstable/using-vala.html . I've greped through the latest 32.1 tarball and there's no indications of libsecret or lsecret so it's probably true for the recent versions as well.

A greater concern is that libsecret is designed to replace gnome-keyring. And I'm concerned compiling vala with libsecret support will break the gnome-keyring support... If that the case, it will be necessary to add an optional argument for libsecret support including a patch and the build input that defaults to false.

**Comment by abbradar** [comment]
I actually meant it other way around -- adding `vala` as a dependency of `libsecret`.

**Comment by RamKromberg** [comment]
> meant it other way around

What I'm saying is that it's going to be a circular dependency. Vala and libsecret will each need an unwrapped version, and then compile against each-other. Simply adding vala to libsecret will work now. But in a couple of month when libsecret stabilizes, it's going to be messy.

**Comment by abbradar** [comment]
Hm, I'm unsure why `vala` would need `libsecret` as an input at all. What I imagine is:

* `vala` (the language)
* `libsecret` (the library, needs `vala` for bindings)
* `dejadup` (the application, needs both of them)

To tell you the truth, I'm not really familiar with Vala or its NixOS infrastructure so I can be just plain wrong with my assumptions.

**Comment by RamKromberg** [comment]
I'm not too familiar with either myself. A while back I added the vala support to libsoup so I vicariously picked up a few tidbits of this and that... But I'm really not too sure how it all adds up. 

I'm currently trying to update vala to 32.1 so I'll push a patch with an optional argument so we'll have something more concrete to mess around with at least.

**Comment by RamKromberg** [comment]
Ok. So it looks like all my concerns & assumptions were outdated or just wrong. There's built-in vala support for gnome-keyring (https://github.com/GNOME/vala/blob/master/vapi/packages/gnome-keyring-1/gnome-keyring-1-custom.vala) and at some point there was libsecret commits which created my confusion. But right now it seems mostly separated so there shouldn't be circular dependencies...
I'm still trying to figure it all out, but I already pushed the vala update here https://github.com/NixOS/nixpkgs/pull/17865 without libsecret since that much should be fine.
I'll try doing something about libsecret next.

**Comment by RamKromberg** [comment]
ok. Like you said, all the libsecret package needed was a simple vala dependency. HOWEVER, the vala default package need to change to the recent vala over runtime bugs (related to the issues I talked about before that were sorted out in upstream).

**Comment by abbradar** [comment]
Awesome! Thank you for sorting this out. I still don't understand much about Vala (why would one ember support for library like libsecret or gnome-keyring into a programming language?), but well, at least it seems we got things on the right side now ^_^

**Comment by RamKromberg** [comment]
> why would one ember support for library like libsecret or gnome-keyring into a programming language?

One problem is that vala uses gobject's types and compiles to C so it's joined at the hip to glib among other things.

Specifically to the bindings, it was done partially for convenience since it started purely as a GNOME domain specific language. And partially because the guys at GNOME pushing vala couldn't get every project and library to maintain their bindings: https://wiki.gnome.org/Projects/Vala/UpstreamGuide

Anyhow, it doesn't really matter now that it all sorted out. So, good luck with the commit. :)

**Comment by s1lvester** [comment]
I made a little progress in [this commit](https://github.com/NixOS/nixpkgs/pull/17844/commits/da55e074c10cb0ba6a0065ba3f45ad3183440ea9) but now I'm stuck with another strange issue:

```bash
help2man: can't get `--help' info from /tmp/nix-build-dejadup-34.2.drv-0/deja-dup-34.2/build/deja-dup/deja-dup
Try `--no-discard-stderr' if option outputs to stderr

```

**Comment by RamKromberg** [comment]
This will get you compiled with the docs:
```
{ stdenv, fetchurl, cmake, pkgconfig, vala_0_32, gettext
, gnome3, libnotify, intltool, itstool, glib, gtk3
, packagekit, libsecret, pcre, libxkbcommon, wrapGAppsHook
, libpthreadstubs, libXdmcp, epoxy, at_spi2_core, dbus
, docSupport ? true, help2man ? null
}:
assert docSupport -> help2man != null;

stdenv.mkDerivation rec {
  name = "dejadup-${version}";
  version = "34.2";

  src = fetchurl {
    url = "https://launchpad.net/deja-dup/34/34.2/+download/deja-dup-34.2.tar.xz";
    sha256 = "ff30732f28233e2422d069f1769a725d4a39c91a279c3bf885c4162fd85818bb";
  };

  prePatch = ''
    #Disable help2man and tests
    substituteInPlace CMakeLists.txt \
      --replace "add_subdirectory(man)" "#add_subdirectory(man)" \
      --replace "add_subdirectory(tests)" "#add_subdirectory(tests)"
  '' + stdenv.lib.optionalString docSupport ''
    #Use Nix's help2man
    rm man/help2man
    sed -i 's,{CMAKE_CURRENT_SOURCE_DIR}/help2man,()${help2man}/bin/help2man,g' man/CMakeLists.txt
  '';

  postInstall = stdenv.lib.optionalString docSupport ''
    #2nd pass for help2man
    unset wrapGAppsHookHasRun
    cd ..
    sed -i 's,{CMAKE_BINARY_DIR}/deja-dup/preferences,(out)/bin,g' man/CMakeLists.txt
    sed -i 's,{CMAKE_BINARY_DIR}/deja-dup,(out)/bin,g' man/CMakeLists.txt
    sed -i 's,-L en_US.UTF-8,,g' man/CMakeLists.txt
    substituteInPlace CMakeLists.txt \
      --replace "#add_subdirectory(man)" "add_subdirectory(man)"
    cmake -DCMAKE_INSTALL_PREFIX:PATH=$out .
    make
    rm -rf $out/*
    make install
  '';

  # couldn't find gio/gdesktopappinfo.h
  NIX_CFLAGS_COMPILE = "-I${glib.dev}/include/gio-unix-2.0";

  buildInputs = [
   cmake pkgconfig vala_0_32 gettext libnotify gnome3.libpeas
   intltool itstool glib gtk3 packagekit libsecret help2man
   pcre libxkbcommon libpthreadstubs libXdmcp epoxy
   at_spi2_core dbus wrapGAppsHook help2man
  ];

  meta = with stdenv.lib; {
    description = "A simple backup tool";
    longDescription = ''
      Déjà Dup is a simple backup tool. It hides the complexity \
      of backing up the Right Way (encrypted, off-site, and regular) \
      and uses duplicity as the backend.
    '';
    homepage = https://launchpad.net/deja-dup;
    license = with licenses; gpl3;
    maintainers = with maintainers; [ ];
    platforms = with platforms; linux;
  };
}
```

You'll also want to write the module for the service; And maybe move some stuff to native inputs... Probably propagate dbus as well...

Good luck :)

**Comment by s1lvester** [comment]
F*ck. I got myself in way deeper waters than I'm able to navigate. After reading up on all the dependencies of gnome-system-services I think I'll have to give up on this one...

**Comment by RamKromberg** [comment]
Aww don't give up :/
Maybe just squash it all up to a commit without a module and keep an issue open for the missing module?
Having the package pull-able means people can write the module by importing a small .nix into their _configuration.nix_ instead of building a whole VM \ container tree for it so it won't go to waste...

**Comment by s1lvester** [comment]
@RamKromberg alright, alright. I'll give it another try as soon as I find some time. After all my laptop doesn't do any backups right now :open_mouth: 
I'll ping you once I got something new.

**Comment by RamKromberg** [comment]
lol good man :+1: 
If you want to backup a few directories, try syncthing. Works nicely enough and very well supported in Nix.

**Pull request closed without merging**